sstables/sstables.cc

```diff
@@ -2678,21 +2678,21 @@ fsync_directory(const io_error_handler& error_handler, sstring fname) {
 
 static future<>
 remove_by_toc_name(sstring sstable_toc_name) {
-    auto dir = dirname(sstable_toc_name);
     sstring prefix = sstable_toc_name.substr(0, sstable_toc_name.size() - sstable_version_constants::TOC_SUFFIX.size());
     sstring new_toc_name = prefix + sstable_version_constants::TEMPORARY_TOC_SUFFIX;
 
     sstlog.debug("Removing by TOC name: {}", sstable_toc_name);
     if (co_await sstable_io_check(sstable_write_error_handler, file_exists, sstable_toc_name)) {
         // If new_toc_name exists it will be atomically replaced.  See rename(2)
         co_await sstable_io_check(sstable_write_error_handler, rename_file, sstable_toc_name, new_toc_name);
-        co_await fsync_directory(sstable_write_error_handler, dir);
+        co_await fsync_directory(sstable_write_error_handler, new_toc_name);
```

Comment:
wouldn't that call fsync on the file rather than the directory?
I think this should be:
```
co_await fsync_directory(sstable_write_error_handler, dirname(sstable_toc_name));
```

Reply:
fsync_directory() calls dirname() on file param. I added a sentence in the commit msg explaining that. So I think it's correct to call fsync_directory() with new_toc_name, which all path components do exist.

Reply:
fsync_directory() is defined like this:
```
future<>
fsync_directory(const io_error_handler& error_handler, sstring fname)
```

then its open dir like this:
`        return open_checked_directory(error_handler, dirname(fname)).then([] (file f) {`

Reply:
ah, ok, then it will work then. Sorry about the noise

Reply:
So it's not just that the old code sometimes passed the wrong thing, it always passed the wrong thing because it passed the parent directory instead of a file in that directory as expected by the fsync_directory function?
By the way, fsync_directory() should be renamed fsync_parent() - it's too confusing :-(

Reply:
> So it's not just that the old code sometimes passed the wrong thing, it always passed the wrong thing because it passed the parent directory instead of a file in that directory as expected by the fsync_directory function? 

Hi @nyh,

Prior to this patch, if remove_by_toc_name() were called on a SSTable containing only TMP TOC

an exception would be triggered in this line: https://github.com/scylladb/scylla/blob/bce2933d99d5e5b86fb0cfccb2882f0b1710db8d/sstables/sstables.cc#L2681

because TOC doesn't exist, meaning fsync_directory() wouldn't be called.

If TOC exists, for a sealed SSTable, then fsync_directory() is called on the directory which the SSTable lives in. Which is exactly what we want to guarantee filesystem has properly updated directory metadata.

> By the way, fsync_directory() should be renamed fsync_parent() - it's too confusing :-(

If I pass /foo/bar, where bar is a directory, I expect fsync_directory() to run on /foo/bar, not on /foo.

If I instead pass /foo/bar, where bar is a regular file, it still makes sense that fsync_directory() picks the underlying directory, which is /foo.

Which makes me think fsync_directory() naming is good as-is.

After my patch, fsync_directory() is now called with absolute path to TMP TOC, meaning we still want it to sync the underlying directory, the directory which TMP TOC lives in.

Reply:
> So it's not just that the old code sometimes passed the wrong thing, it always passed the wrong thing because it passed the parent directory instead of a file in that directory as expected by the fsync_directory function?

Right.

> By the way, fsync_directory() should be renamed fsync_parent() - it's too confusing :-(

I agree with @raphaelsc that it can be fixed to what its name suggests it does (and documenting that) by fstat'ing `fname` to determine if it's a directory or not and fsync'ing its parent directory only if it is not a directory by itself.

But I'm not sure it's worth the extra logic, since the caller knows what it wants to fsync and can pass the actual directory to fsync.  Note that we're still doing the wrong thing at the bottom of this function:
https://github.com/scylladb/scylla/blob/bcd35af7cf93eac9664a044816847521cfddb0ac/sstables/sstables.cc#L2731

Reply:
> If TOC exists, for a sealed SSTable, then fsync_directory() is called on the directory which the SSTable lives in. Which is exactly what we want to guarantee filesystem has properly updated directory metadata.

My comment here was about the **old** (before patch) code being even more wrong that we thought it is (unless I misunderstood something): The old code had:
```c++
auto dir = dirname(sstable_toc_name);
co_await fsync_directory(sstable_write_error_handler, dir);
```

If I understand correctly, if sstable_toc_name is "/a/b/c.toc", then "dir" is "/a/b" and fsync_directory(dir) takes dirname(dir) and gets "/a" and then does a fsync on this "/a". Wasn't that completely wrong (regardless of whether we had toc or toc.tmp)?

Reply:
@nyh I agree with you.  The old implementation was wrong and the current version of this PR is still wrong.
I suggest to fix that as part of this PR by taking the `dirname` call out of `fsync_directory` so its caller will make sure to call it with the right directory to fsync.  There are only 2 call sites: one after the rename of TOC to TemporaryTOC and a final one after we're done removing all components.